consensus/core/src/commit_syncer.rs

```diff
@@ -0,0 +1,586 @@
+
+    pub(crate) async fn stop(self) {
+        let _ = self.tx_shutdown.send(());
+        // Do not abort schedule task, which waits for fetches to shut down.
+        let _ = self.schedule_task.await;
+    }
+
+    async fn schedule_loop(
+        inner: Arc<Inner<C>>,
+        fetch_state: Arc<Mutex<FetchState>>,
+        mut rx_shutdown: oneshot::Receiver<()>,
+    ) {
+        let mut interval = tokio::time::interval(Duration::from_secs(2));
+        interval.set_missed_tick_behavior(MissedTickBehavior::Skip);
+        // Inflight requests to fetch commits from different authorities.
+        let mut inflight_fetches = JoinSet::new();
+        // Additional ranges (inclusive start and end) of commits to fetch.
+        let mut pending_fetches = VecDeque::<(CommitIndex, CommitIndex)>::new();
+        // Highest end index among inflight and pending fetches.
+        let mut highest_scheduled_index = Option::<CommitIndex>::None;
+        // Fetched commits and blocks by commit indices.
+        let mut fetched_blocks = BTreeMap::<(CommitIndex, CommitIndex), Vec<VerifiedBlock>>::new();
+        // The commit index that is the max of local last commit index and highest commit index of blocks sent to Core.
+        let mut synced_commit_index = inner.dag_state.read().last_commit_index();
+
+        loop {
+            tokio::select! {
+                // Periodially, schedule new fetches if the node is falling behind.
+                _ = interval.tick() => {
+                    let highest_valid_index = inner.highest_commit_monitor.highest_valid_index(&inner.context);
+                    // Update synced_commit_index periodically to make sure it is not smaller than
+                    // local commit index.
+                    synced_commit_index = synced_commit_index.max(inner.dag_state.read().last_commit_index());
+                    // TODO: cleanup inflight fetches that are no longer needed.
+                    let fetch_after_index = synced_commit_index.max(highest_scheduled_index.unwrap_or(0));
+                    info!(
+                        "Checking to schedule fetches: synced_commit_index={}, fetch_after_index={}, highest_valid_index={}",
+                        synced_commit_index, fetch_after_index, highest_valid_index
+                    );
+                    // When the node is falling behind, schedule pending fetches which will be executed on later.
+                    'pending: for prev_end in (fetch_after_index..=highest_valid_index).step_by(inner.context.parameters.commit_sync_batch_size as usize) {
+                        // Create range with inclusive start and end.
+                        let range_start = prev_end + 1;
+                        let range_end = prev_end + inner.context.parameters.commit_sync_batch_size;
+                        // Do not fetch with small batches. Block subscription and synchronization will help the node catchup.
+                        if range_end > highest_valid_index {
```

Comment:
Yes this is intentional, to avoid generating a lot of small batches to fetch, which will burden this node and its peers. Suppose the node falls behind by 1 commit every now and then, if the smaller batch that includes the highest_valid_index gets generated, there will be a lot of 1 commit fetches. The existing logic will wait for the node to fall behind by 100 commit, then create a bigger batch. Expanded the comment.